adapters/waardex/waardex.go

```diff
@@ -0,0 +1,288 @@
+	errs := make([]error, 0, len(request.Imp))
+	if len(request.Imp) == 0 {
+		errs = append(errs, newBadInputError("No impression in the bid request"))
+		return nil, errs
+	}
+	imps, impExts, err := getImpressionsInfo(request.Imp)
+	if len(imps) == 0 {
+		return nil, err
+	}
+	errs = append(errs, err...)
+
+	pub2impressions, dispErrors := dispatchImpressions(imps, impExts)
+	if len(dispErrors) > 0 {
+		errs = append(errs, dispErrors...)
+	}
+	if len(pub2impressions) == 0 {
+		return nil, errs
+	}
+	result := make([]*adapters.RequestData, 0, len(pub2impressions))
+	for k, imps := range pub2impressions {
+		bidRequest, err := adapter.buildAdapterRequest(request, &k, imps)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			result = append(result, bidRequest)
+		}
+	}
+	return result, errs
+}
+
+// getImpressionsInfo checks each impression for validity and returns a copy for each impression with corresponding exts
+func getImpressionsInfo(imps []openrtb2.Imp) ([]openrtb2.Imp, []openrtb_ext.ExtImpWaardex, []error) {
+	impsCount := len(imps)
+	errors := make([]error, 0, impsCount)
+	resImps := make([]openrtb2.Imp, 0, impsCount)
+	resImpExts := make([]openrtb_ext.ExtImpWaardex, 0, impsCount)
+
+	for _, imp := range imps {
+		impExt, err := getImpressionExt(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		if err := validateImpression(&imp, impExt); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		resImps = append(resImps, imp)
+		resImpExts = append(resImpExts, *impExt)
+	}
+	return resImps, resImpExts, errors
+}
+
+func validateImpression(imp *openrtb2.Imp, impExt *openrtb_ext.ExtImpWaardex) error {
+	if impExt.ZoneId < 1 {
+		return newBadInputError(fmt.Sprintf("Invalid zoneId value: %d. Ignoring imp id=%s", impExt.ZoneId, imp.ID))
+	}
+	return nil
+}
+
+// Group impressions by Waardex-specific parameter `zoneId`
+func dispatchImpressions(imps []openrtb2.Imp, impsExt []openrtb_ext.ExtImpWaardex) (map[openrtb_ext.ExtImpWaardex][]openrtb2.Imp, []error) {
+	res := make(map[openrtb_ext.ExtImpWaardex][]openrtb2.Imp)
+	errors := make([]error, 0)
+	for idx := range imps {
+		imp := imps[idx]
+		imp.Ext = nil
+		impExt := impsExt[idx]
+		if res[impExt] == nil {
+			res[impExt] = make([]openrtb2.Imp, 0)
+		}
+		if isMultiFormatImp(&imp) {
+			splImps := splitMultiFormatImp(&imp)
+			res[impExt] = append(res[impExt], splImps...)
+		} else {
+			res[impExt] = append(res[impExt], imp)
+		}
+	}
+	return res, errors
+}
+
+func isMultiFormatImp(imp *openrtb2.Imp) bool {
+	count := 0
+	if imp.Video != nil {
+		count++
+	}
+	if imp.Audio != nil {
+		count++
+	}
+	if imp.Banner != nil {
+		count++
+	}
+	if imp.Native != nil {
+		count++
+	}
+	return count > 1
+}
+
+func splitMultiFormatImp(imp *openrtb2.Imp) []openrtb2.Imp {
+	splitImps := make([]openrtb2.Imp, 0, 4)
```

Comment:
Max you will create 2 impressions in this functionality so capacity value `2` value is sufficient instead of `4`